### ISSUE TITLE
Automatic expert QLoRA using chat history reflection and automatic training data collection

### ISSUE THREAD

**Issue by MB-Finski** [microsoft/autogen#474]
So the core idea here is to have each individual expert in MoE to autonomously gather training data and fine-tune itself on that data. 

The proposed solution would contain each "expert" in an MoE architercture in a wrapper class that records the chat history and then periodically reviews the chat history using a "reviewer" expert llm that tries to find (based on the full chat history i.e. also conversation that came nefore/after the original query) improved initial answers to queries that the model received. If it does come up with an improved answer, this query + answer pair is saved as training data. Then QLoRA is run periodically to have an improved fine-tuning adapter for the expert. I think this strategy would be especially useful in the case of RAG augmented experts since the information and 'mental models' contained within the retrieval dataset would slowly over time be built into the expert models.

Ideally all experts, including the reviewer, would have their own wrapper so all parts of the network would improve over time.

I find this idea fascinating especially since this would be very similar to human cognition: i.e. during day time gather experiences and have limited adaptation capability to new topics (i.e. RAG using chat history or some other strategy) then during "sleep" reflect on the chat history and, if necessary, perform a fine-tuning on the relfections. Thismight enable the models to learn completely new ways to think and would slowly over time build 

Do you know of implementations similar to this existing? I was thinking of implementing this type of wrapper class myself but unfortunately I'm bogged down with other work right now so I can't contribute it.

**Comment by ddwinhzy** [comment]
Shouldn't each agent have the autonomy to decide whether or not to collect?

**Comment by MB-Finski** [comment]
Yeah, that should also work. But, then, I'm a bit worried about 'drift'. I.e. since the agents don't gather training data on self-reflection (but rather answering the queries they get) they may slowly degrade in their reflection ability and start generating worse and worse training data.

A separate "reflection" or "reviewer" agent would provide a "ground truth" of what the goal in answering each query should be; it would provide training data with consistent quality over time. Another consideration is that you could then even wrap the reviewer in a self-training wrapper and have it learn to reflect better over time (although this would require some sort of MoE setup such that the reviewer would get feedback on it's actions in chat format).

**Comment by sonichi** [comment]
This sounds a natural extension of the EcoAssistant work done by @JieyuZ2 et al.

**Comment by ddwinhzy** [comment]
It seems like everyone's been waiting for this moment

**Comment by MB-Finski** [comment]
> This sounds a natural extension of the EcoAssistant work done by @JieyuZ2 et al.

Thanks for pointing this out! Combining these strategies could prove to be effective at transferring the performance/knowledge of the large/expensive foundation models to these smaller models used in MoEs.

I so wish I could work on something like this right now -- alas, that is not the case.

**Comment by thinkall** [comment]
Hi @MB-Finski , have you got something workable for this?

**Comment by MB-Finski** [comment]
@thinkall Unfortunately, no. I pitched this idea while I was at Nextcloud but it was, understandably, a bit too experimental to align with the more immediate development needs. Right now I'm back at my day job as an anesthesiologist so I've had even less time to experiment.

**Comment by thinkall** [comment]
Thank you @MB-Finski . This sounds like a research topic rather than an issue of AutoGen. I'll close the issue and you're very welcome to re-open it or create a new issue/PR once you get some time to work on it.
